internal/webhook/management_webhook.go

```diff
@@ -121,99 +126,71 @@ func checkComponentsRemoval(ctx context.Context, cl client.Client, oldMgmt, newM
 		return nil
 	}
 
-	release := new(kcmv1.Release)
-	if err := cl.Get(ctx, client.ObjectKey{Name: newMgmt.Spec.Release}, release); err != nil {
-		return fmt.Errorf("failed to get Release %s: %w", newMgmt.Spec.Release, err)
-	}
-
-	removedProvidersSet := make(map[string]struct{})
+	var errs error
 	for _, m := range removedComponents {
 		tplRef := m.Template
-		if tplRef == "" {
+		if tplRef == "" && release != nil {
 			tplRef = release.ProviderTemplate(m.Name)
 		}
 
-		// it does not matter if component has been successfully installed
+		if tplRef == "" {
+			continue
+		}
+
 		prTpl := new(kcmv1.ProviderTemplate)
 		if err := cl.Get(ctx, client.ObjectKey{Name: tplRef}, prTpl); err != nil {
 			return fmt.Errorf("failed to get ProviderTemplate %s: %w", tplRef, err)
 		}
 
-		for _, pn := range prTpl.Status.Providers {
-			removedProvidersSet[pn] = struct{}{}
+		inUsedProviders, err := getInUsedProviderWithContracts(ctx, cl, prTpl)
+		if err != nil {
+			return err
 		}
-	}
-
-	if len(removedProvidersSet) == 0 { // sanity
-		return nil
-	}
-
-	for providerName := range removedProvidersSet {
-		clusterTemplates := new(kcmv1.ClusterTemplateList)
-		if err := cl.List(ctx, clusterTemplates, client.MatchingFields{kcmv1.ClusterTemplateProvidersIndexKey: providerName}); err != nil {
-			return fmt.Errorf("failed to list ClusterTemplates: %w", err)
-		}
-
-		if len(clusterTemplates.Items) == 0 {
+		if len(inUsedProviders) == 0 {
 			continue
 		}
-
-		for _, cltpl := range clusterTemplates.Items {
-			mcls := new(kcmv1.ClusterDeploymentList)
-			if err := cl.List(ctx, mcls,
-				client.MatchingFields{kcmv1.ClusterDeploymentTemplateIndexKey: cltpl.Name},
-				client.Limit(1)); err != nil {
-				return fmt.Errorf("failed to list ClusterDeployments: %w", err)
-			}
-
-			if len(mcls.Items) == 0 {
-				continue
-			}
-
-			return fmt.Errorf("provider %s is required by at least one ClusterDeployment (%s) and cannot be removed from the Management %s", providerName, client.ObjectKeyFromObject(&mcls.Items[0]), newMgmt.Name)
+		for provider := range inUsedProviders {
+			errs = errors.Join(errs, fmt.Errorf("provider %s is required by at least one ClusterDeployment and cannot be removed from the Management %s", provider, newMgmt.Name))
```

Comment:
not sure if it is important: in theory, one could remove several providers from the `spec` that have intersections in terms of bootstrap/cp/infra providers and the errors might contain dups, hence it might be beneficial to put providers into a map and only then concat the errors